ambry-router/src/test/java/com.github.ambry.router/MockServer.java

```diff
@@ -53,54 +73,139 @@
-        final ServerErrorCode putError =
-            hardError != null ? hardError : putErrors.size() > 0 ? putErrors.poll() : ServerErrorCode.No_Error;
-        if (putError == ServerErrorCode.No_Error) {
-          updateBlobMap(putRequest);
-        }
-        return new MockBoundedByteBufferReceive(
-            new PutResponse(putRequest.getCorrelationId(), putRequest.getClientId(), putError) {
-              ByteBuffer getPayload()
-                  throws IOException {
-                ByteArrayOutputStream bStream = new ByteArrayOutputStream();
-                DataOutputStream dStream = new DataOutputStream(bStream);
-                dStream.writeShort((short) RequestOrResponseType.PutResponse.ordinal());
-                dStream.writeShort(versionId);
-                dStream.writeInt(correlationId);
-                dStream.writeInt(0); // avoiding adding clientId
-                dStream.writeShort((short) putError.ordinal());
-                return ByteBuffer.wrap(bStream.toByteArray());
-              }
-            }.getPayload());
-
+        response = makePutResponse((PutRequest) send, serverError);
+        break;
+      case GetRequest:
+        response = makeGetResponse((GetRequest) send, serverError);
+        break;
       case DeleteRequest:
-        final DeleteRequest deleteRequest = (DeleteRequest) send;
-        final String blobIdString = deleteRequest.getBlobId().getID();
-        final ServerErrorCode error = getErrorFromBlobIdStr(blobIdString);
-        return new MockBoundedByteBufferReceive(
-            new DeleteResponse(deleteRequest.getCorrelationId(), deleteRequest.getClientId(), error) {
-              ByteBuffer getPayload()
-                  throws IOException {
-                ByteArrayOutputStream bStream = new ByteArrayOutputStream();
-                DataOutputStream dStream = new DataOutputStream(bStream);
-                dStream.writeShort((short) RequestOrResponseType.DeleteResponse.ordinal());
-                dStream.writeShort(versionId);
-                dStream.writeInt(correlationId);
-                dStream.writeInt(0); // avoiding adding clientId
-                dStream.writeShort((short) error.ordinal());
-                return ByteBuffer.wrap(bStream.toByteArray());
-              }
-            }.getPayload());
+        response = makeDeleteResponse((DeleteRequest) send, serverError);
+        break;
       default:
         throw new IOException("Unknown request type received");
     }
+    ByteBufferChannel channel = new ByteBufferChannel(ByteBuffer.allocate((int) response.sizeInBytes()));
+    response.writeTo(channel);
+    ByteBuffer payload = channel.getBuffer();
+    payload.flip();
+    // read the size off. the size is used by the client to allocate the buffer and then interpret the response.
+    // MockServer abstracts it away at a level above that and returns the "allocated buffer".
+    payload.getLong();
+    return new MockBoundedByteBufferReceive(payload);
```

Comment:
as per our discussion, `MockBoundedByteBufferReceive` might not be required and using `BoundedByteBufferReceive` will make line 102 irrelevant.